libraries/AP_Mission/AP_Mission.cpp

```diff
@@ -1864,6 +1879,133 @@ bool AP_Mission::jump_to_abort_landing_sequence(void)
     return false;
 }
 
+// check which is the shortest route to landing an RTL via a DO_LAND_START or continuing on the current mission plan
+bool AP_Mission::is_best_land_sequence(void)
+{
+    // check if there is even a running mission to interupt
+    if (_flags.state != MISSION_RUNNING) {
+        return false;
+    }
+
+    // check if aircraft has already jumped to a landing sequence
+    if (_flags.in_landing_sequence) {
+        return true;
+    }
+
+    // check if MIS_OPTIONS bit set to allow distance calculation to be done
+    if (!(_options & AP_MISSION_MASK_DIST_TO_LAND_CALC)) {
+        return false;
+    }
+
+    // The decision to allow a failsafe to interupt a potential landing approach 
+    // is a distance travelled minimization problem.  Look forward in 
+    // mission to evaluate the shortest remaining distance to land.
+
+    // go through the mission for the nearest DO_LAND_START first as this is the most probable route 
+    // to a landing with the minimum number of WP.
+    uint16_t do_land_start_index = get_landing_sequence_start();
+    if (do_land_start_index == 0) {
+        // then no DO_LAND_START commands are in mission and normal failsafe behaviour should be maintained
+        return false;
+    }
+
+    // get our current location
+    Location current_loc;
+    if (!AP::ahrs().get_position(current_loc)) {
+        // we don't know where we are!!
+        return false;
+    }
+
+    // get distance to landing if travelled to nearest DO_LAND_START via RTL
+    float dist_via_do_land;
+    if (!distance_to_landing(do_land_start_index, dist_via_do_land, current_loc)) {
+        // cant get a valid distance to landing
+        return false;
+    }
+
+    // get distance to landing if continue along current mission path
+    float dist_continue_to_land;
+    if (!distance_to_landing(_nav_cmd.index, dist_continue_to_land, current_loc)) {
+        // cant get a valid distance to landing
+        return false;
+    }
+
+    // compare distances
+    if (dist_via_do_land >= dist_continue_to_land) {
+        // then the mission should carry on uninterrupted as that is the shorter distance
+        gcs().send_text(MAV_SEVERITY_NOTICE, "Rejecting RTL: closer land if mis continued");
+        _flags.in_landing_sequence = true;
```

Comment:
After more consideration, why do you even need to flag this? This was requested to figure out if we should switch to it. If it was true and we haven't jumped off somehow (IE by the GCS) then it would be true if we recomputed it, which realistically we wouldn't anyways unless another failsafe hit).

Reply:
This one is slightly misleading  grant you, however I found that there are certain instances when the distance calculation is called twice.  As it is deemed an expensive calculation, setting the flag to true prevents this.  I believe that the case when the function is called twice is when the failsafe is action is to land but the distance to landing calculation leads to the switch statement in events.cpp to fall through.  So the function is called again in RTL.

Reply:
Done as suggested.  Fixed the original issue in events.cpp which caused the fall through to RTL when the aircraft is already on the shortest distance to landing.  I have also done what you suggested and taken the changing flag state out of the is_ function.  The flag is now set from events.cpp